Introduce Apply plan node for subqueries

kokosing opened the pull request:
It is a logical representation of a subquery.
    Apply node contains input relation, expression relation and parameters
    (symbol list).
    
    Each row in the input relation is passed to the expression, then
    expression (subquery) is executed. Expression produces new columns which
    are joined with the input row. Parameters (correlation list) determines
    what columns need to be passed from the input to the expression to
    calculate the result rows. If correlation list is empty then subquery
    is not correlated, in this case values returned from expression are
    always the same, not dependent on input relation.
    
    See
    https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:
```diff
@@ -634,41 +633,34 @@ private PlanBuilder appendProjections(PlanBuilder subPlan, Iterable<Expression>
     }
 
-    private PlanBuilder appendSemiJoin(PlanBuilder subPlan, InPredicate inPredicate)
+    private PlanBuilder appendInPredicateApply(PlanBuilder subPlan, InPredicate inPredicate)
```
this code is identical with the one in QueryPlanner? Should I refactor it?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java:
```diff
@@ -266,8 +265,8 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
 
             // Add semi joins if necessary
             if (joinInPredicates != null) {
```
What are the joinInPredicates?

cberner commented:
Looks like the tests are failing. Can you take a look at Travis?

pnowojski commented:
@cberner @kokosing: travis fixed, top commit should to be squashed before merging with master.

kokosing commented:
@cberner @martint ping

cberner commented:
I'm working with @sopel39 on `EXPLAIN ANALYZE` at the moment. Can you remind me once that is done?

kokosing commented:
@cberner @martint ping

kokosing commented:
@martint Can you please review? This is base refactor needed to start the work on correlated subqueries.

martint commented on presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:
```diff
@@ -334,6 +336,9 @@ protected void setup(Binder binder)
         binder.bind(new TypeLiteral<List<PlanOptimizer>>() {}).toProvider(PlanOptimizersFactory.class).in(Scopes.SINGLETON);
 
+        // plan sanity checkers
+        binder.bind(new TypeLiteral<List<PlanSanityChecker>>() {}).toProvider(PlanSanityCheckersFactory.class).in(Scopes.SINGLETON);
```
I don't think this abstraction is necessary, especially since they don't have dependencies on other classes (like metadata, etc). I would just handle all this within PlanSanityChecker (which could hold a list of checkers internally and iterate over them in the `validate()` call`

kokosing commented:
@martint comment applied and rebased

martint commented:
I merged the first 5 commits.

kokosing commented:
Thanks. I rebased.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoSubqueryExpressionLeftSanityChecker.java:
```diff
@@ -0,0 +1,77 @@
+                @Override
+                protected Void visitSubqueryExpression(SubqueryExpression node, Void context)
+                {
+                    throw new SemanticException(SemanticErrorCode.NOT_SUPPORTED, node, "Given subquery is not yet supported");
```
Plan sanity checkers are about enforcing invariants. I.e., make sure optimizers are not breaking assumptions and producing incorrect plans. So, they should throw IllegalArgumentExceptions instead of SemanticExceptions (the analyzer should be throwing these)

martint commented on presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:
```diff
@@ -156,10 +156,10 @@ protected void setup(Binder binder)
         if (serverConfig.isCoordinator()) {
             discoveryBinder(binder).bindHttpAnnouncement("presto-coordinator");
-            binder.bind(new TypeLiteral<Optional<QueryPerformanceFetcher>>(){}).toProvider(QueryPerformanceFetcherProvider.class).in(Scopes.SINGLETON);
+            binder.bind(new TypeLiteral<Optional<QueryPerformanceFetcher>>() {}).toProvider(QueryPerformanceFetcherProvider.class).in(Scopes.SINGLETON);
```
Unrelated changes

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/EnsureSymbolDependenciesPlanSanityChecker.java:
```diff
@@ -63,11 +65,11 @@
  * Ensures that all dependencies (i.e., symbols in expressions) for a plan node are provided by its source nodes
  */
-public final class PlanSanityChecker
+public final class EnsureSymbolDependenciesPlanSanityChecker
```
Call this "ValidateDependencies"

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java:
```diff
@@ -0,0 +1,98 @@
+    private final Map<Symbol, Symbol> correlation;
+
+    @JsonCreator
+    public ApplyNode(@JsonProperty("id") PlanNodeId id,
```
Move argument to the next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java:
```diff
@@ -0,0 +1,98 @@
+    private final PlanNode subquery;
+
+    // Mapping between input and subquery, key represents symbols in input, values in subquery
+    private final Map<Symbol, Symbol> correlation;
```
I'm not sure I understand how this is supposed to work. The thing that confuses me is that it's linking the fields from the outer subplan to the outputs of the inner subplan, but in my mental model, the fields in the outer are supposed to be *inputs* to the inner that get "bound" on every iteration of the inner subquery (i.e., like arguments to a function call).

kokosing replied:
I think your mental model is correct at least it is in line with my mental model. Maybe I didn't cast that into java code clearly enough. I think that comment above correlation map could be misleading, I will fix it.

The direction of mapping in correlation map to me does not have much importance. Would it be cleaner to you I would reverse it? In that case, key in map would represent some symbol in a subquery plan (inner plan) - an input to subquery call, and value in map would represent output symbol from input node (outer plan).

Take a look at the example of a real query:
`SELECT n.nationkey FROM nation n WHERE n.nationkey + 100 < (SELECT n.nationkey*2);`

This query could have below plan:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) {nationkey: nationkey_1}
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey_1 * 2) -> (expr)
```
At the above example, `input` of apply node returns `nationkey` symbol which is then used in `subquery` plan in place of `nationkey_1` symbol.

However, I have an idea of using plain list of correlation symbols instead of map. Then plan would like below:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) [nationkey]
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey * 2) -> (expr)
```

See that input and subquery plans are both using the same symbol `nationkey` (no need for `nationkey_1`). So Apply need to only know which symbols are correlated. I think it is easier to understand such plan. I will change that.

martint replied:
Ok, that makes sense. But, then, isn't the following check in the `ApplyNode` constructor incorrect?

```java
checkArgument(subquery.getOutputSymbols().containsAll(correlation.values()), "Subquery does not contain symbols from correlation");
```

For your first example above, this would be checking that `expr` (output of the inner subquery) is in the values of the correlation map, which contains `nationkey`->`nationkey_1`

kokosing replied:
Yes, it is incorrect. I noticed that and I removed it in subsequent commit. Please see: https://github.com/prestodb/presto/pull/4403/commits/d15cf99ca337665758cae48f50c952a85cf198f2#diff-727404f7967058fb13c2898cad1c3784R52

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanSanityChecker.java:
```diff
@@ -0,0 +1,38 @@
+ */
+public interface PlanSanityChecker
+{
+    List<PlanSanityChecker> PLAN_SANITY_CHECKERS = ImmutableList.of(
```
I think it's confusing to have this on the interface that the checkers implement. I'd have a PlanSanityChecker class with a single static method (and a static list of checkers), and a separate interface checkers implement. E.g.:

```java
class PlanSanityChecker
{
    public static void validate(PlanNode plan) { ... }

    public interface Checker 
    {
        void validate(PlanNode plan);
    }
}

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java:
```diff
@@ -0,0 +1,70 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class UncorrelatedScalarApplyRemover
```
I'd might rename this to TransformUncorrelatedScalarToJoin

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java:
```diff
@@ -0,0 +1,70 @@
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            node = (ApplyNode) super.visitApply(node, context);
```
Call `context.defaultRewrite(...)` instead.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java:
```diff
@@ -0,0 +1,70 @@
+
+            if (node.getCorrelation().isEmpty() && node.getSubquery() instanceof EnforceSingleRowNode) {
+                // only scalar subquery wraps expression in EnforceSingleRowNode
+                return new JoinNode(idAllocator.getNextId(),
```
Move argument to next line

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.Map;
+
+public class VerifyNoApplyNodeLeftPlanOptimizer
```
I'm not sure I like the idea of having this as a plan optimizer. Can we make it a sanity checker, instead?

kokosing replied:
It used to be written as sanity checker. 

Plan sanity checkers are run before optimizers where apply nodes still exist. I could change sanity checkers to accept parameter flag which would tell if plan is optimized or not, but to make this work all other optimizers would need to support Apply node. Otherwise, instead of getting an error from sanity checkers we could get some mysterious error from other optimizer rule.

I don't like it either but I don't have a better idea for this.

martint replied:
We can just remove the call to validate the plan before the optimizers run.

kokosing replied:
Well... I think that having sanity checks before optimizers run has its own value. Thanks to that you know if plan was broken before or after optimizers run.

I would prefer to leave it as it is, unless you have a strong opinion about it. Then I will change it according to your comment.

martint replied:
Generally, the way this ends up being used is:

* Get a "broken plan error"
* Distill the query to a small reproducible case
* Fire up Presto in a debugger & trace. This commonly involves adding a call to validate() before/after each optimizer.

So, you almost always have to tweak the code to use track down issues, so I think it's acceptable to just validate at the end and leave it up to the developer to insert calls and enable/disable checkers as necessary while debugging.

kokosing replied:
Sounds ok. Thanks for explanation.

śr., 4.05.2016, 21:46 użytkownik Martin Traverso <notifications@github.com>
napisał:

> In
> presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java
> <https://github.com/prestodb/presto/pull/4403#discussion_r62102119>:
>
> > +package com.facebook.presto.sql.planner.optimizations;
> > +
> > +import com.facebook.presto.Session;
> > +import com.facebook.presto.spi.PrestoException;
> > +import com.facebook.presto.spi.StandardErrorCode;
> > +import com.facebook.presto.spi.type.Type;
> > +import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
> > +import com.facebook.presto.sql.planner.SimplePlanVisitor;
> > +import com.facebook.presto.sql.planner.Symbol;
> > +import com.facebook.presto.sql.planner.SymbolAllocator;
> > +import com.facebook.presto.sql.planner.plan.ApplyNode;
> > +import com.facebook.presto.sql.planner.plan.PlanNode;
> > +
> > +import java.util.Map;
> > +
> > +public class VerifyNoApplyNodeLeftPlanOptimizer
>
> Generally, the way this ends up being used is:
>
>    - Get a "broken plan error"
>    - Distill the query to a small reproducible case
>    - Fire up Presto in a debugger & trace. This commonly involves adding
>    a call to validate() before/after each optimizer.
>
> So, you almost always have to tweak the code to use track down issues, so
> I think it's acceptable to just validate at the end and leave it up to the
> developer to insert calls and enable/disable checkers as necessary while
> debugging.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/prestodb/presto/pull/4403/files/745f4349fafc47aa5051fb3a577af7332ddd11ef#r62102119>
>

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java:
```diff
@@ -0,0 +1,281 @@
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            node = (FilterNode) super.visitFilter(node, context);
```
Call `context.defaultRewrite(...)` instead

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java:
```diff
@@ -0,0 +1,281 @@
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            node = (ProjectNode) super.visitProject(node, context);
```
Same here

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java:
```diff
@@ -0,0 +1,281 @@
+            return assignmentsBuilder.build();
+        }
+
+        private List<InPredicate> collectInPredicates(Collection<Expression> expressions)
```
I'd call this method `extractInPredicates`

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java:
```diff
@@ -0,0 +1,281 @@
+                new DefaultTraversalVisitor<Void, Void>()
+                {
+                    @Override
+                    protected Void visitInPredicate(InPredicate node, Void context)
```
This will also include any predicates of the form `xxx in (a, b, c, ...)`. Is that intentional? Isn't this supposed to do it only for cases where the "value expression" of the IN predicate is a subquery?

kokosing replied:
For cases of regular IN predicate - `xxx in (a, b, c, ...)` optimizer won't find any matching `Apply` node, so it does not influence the correctness (only performance suffers a bit).

At this point subquery expression is rewritten to apply node `xxx in (select...)` is in form of `xxx in expr`. Where `expr` is a symbol referring to apply node output. Therefore it could be difficult to properly distinguish IN predicate with subquery to IN predicate with other symbol `xxx in (yyy)`, where `yyy` could be a reference to a column for example.

However, I could optimize this to select IN predicates which have only one symbol in valueList.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java:
```diff
@@ -0,0 +1,281 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class UncorrelatedInPredicateApplyRemover
```
Can you add an example to javadoc comments of whats kind of plan pattern it applies to and what the result looks like?

kokosing commented:
@martint I addressed all the comments and answered all the questions. Comments are addressed in separate commits, I will merge them just before the merge. I didn't rebase on origin/master because I wanted to preserve the discussion comments for a while.

kokosing commented:
Superseded by #5202
